ambry-store/src/main/java/com/github/ambry/store/BlobStore.java

```diff
@@ -1142,6 +1109,137 @@ public void shutdown() throws StoreException {
     shutdown(false);
   }
 
+  /**
+   * Update the sealed status of the replica.
+   */
+  private void updateSealedStatus() {
+    // In two zk clusters case, "replicaId.isSealed()" might be true in first cluster but is still unsealed in second
+    // cluster, as server only spectates the first cluster. To guarantee both clusters have consistent seal/unseal
+    // state, we bypass "isSealed()" check if there are more than one replicaStatusDelegates.
+    ReplicaSealStatus replicaSealStatus1 = replicaSealStatus.get();
+    if (resolveReplicaSealStatusFromLogSize() == ReplicaSealStatus.SEALED && (replicaSealStatus.get() != ReplicaSealStatus.SEALED || (
+        replicaStatusDelegates.size() > 1 &&  replicaSealStatus.getAndSet(ReplicaSealStatus.SEALED) != ReplicaSealStatus.SEALED))) {
+      for (ReplicaStatusDelegate replicaStatusDelegate : replicaStatusDelegates) {
```

Comment:
Can replicaStatusDelegates have 2 and more entries?
If yes, what if one replicaStatusDelegate.seal is true but the other replicaStatusDelegate.seal is false?

The old code probably has the same code logic. Just want to understand it.

Reply:
So replicaStatusDelegates can have 2 entries during zk migration etc. Out strategy during zk migration was that we will use the replicaStatusDelegates for both the zk endpoints/clusters, and let them sync up for a few days before decommissioning the older zk.

Reply:
So whether a replica is in sealed/partially_sealed/unsealed state is determined by only three factors
1. Whether the size of used capacity of a replica's log exceeds a certain threshold.
2. Whether there is any other replica of the same partition marked as sealed in helix, so the replica gets a notification from helix.
3. Whether the partition id of the replica is present in partitionOverrideList (store in helix)

In all the cases, if there is any divergence, then it's safe for a replica to temporarily assume the most strict seal state. With compaction and writes, each replica will keep re-evaluating its state and eventually come back to a state that is guided by its log usage.

Reply:
In this `BlobStore` logic, the only thing that matters for a replica to mark itself as sealed/partially-sealed/unsealed is the log size alone (factor 1 above).

Reply:
Thanks for the explanation.